net/hclouddns/src/opnsense/scripts/ddclient/lib/account/hetzner_cloud.py

```diff
@@ -0,0 +1,275 @@
+                    self.description, response.status_code, response.text
+                )
+            )
+            return None
+
+        try:
+            payload = response.json()
+        except requests.exceptions.JSONDecodeError:
+            syslog.syslog(
+                syslog.LOG_ERR,
+                "Account %s error parsing JSON response [zones]: %s" % (self.description, response.text)
+            )
+            return None
+
+        zones = payload.get('zones', [])
+        if not zones:
+            syslog.syslog(
+                syslog.LOG_ERR,
+                "Account %s zone '%s' not found" % (self.description, zone_name)
+            )
+            return None
+
+        zone_id = zones[0].get('id')
+        if self.is_verbose:
+            syslog.syslog(
+                syslog.LOG_NOTICE,
+                "Account %s found zone ID %s for %s" % (self.description, zone_id, zone_name)
+            )
+
+        return zone_id
+
+    def _get_record(self, headers, zone_id, record_name, record_type):
+        """Get existing record by name and type"""
+        url = f"{self._api_base}/zones/{zone_id}/rrsets/{record_name}/{record_type}"
+
+        response = requests.get(url, headers=headers)
+
+        if response.status_code == 404:
+            return None
+
+        if response.status_code != 200:
+            syslog.syslog(
+                syslog.LOG_ERR,
+                "Account %s error fetching record: HTTP %d - %s" % (
+                    self.description, response.status_code, response.text
+                )
+            )
+            return None
+
+        try:
+            payload = response.json()
+            return payload.get('rrset')
+        except requests.exceptions.JSONDecodeError:
+            syslog.syslog(
+                syslog.LOG_ERR,
+                "Account %s error parsing JSON response [record]: %s" % (self.description, response.text)
+            )
+            return None
+
+    def _update_record(self, headers, zone_id, record_name, record_type, address):
+        """Update existing record with new address
+
+        NOTE: Hetzner Cloud API has a bug where PUT returns 200 but doesn't update.
+        Workaround: DELETE old record, then POST new record.
+        """
+        # DELETE old record first
+        delete_url = f"{self._api_base}/zones/{zone_id}/rrsets/{record_name}/{record_type}"
+        delete_response = requests.delete(delete_url, headers=headers)
+
+        if delete_response.status_code not in [200, 201, 204]:
+            syslog.syslog(
+                syslog.LOG_ERR,
+                "Account %s error deleting record for update: HTTP %d - %s" % (
+                    self.description, delete_response.status_code, delete_response.text
+                )
+            )
+            return False
+
+        # CREATE new record
+        return self._create_record(headers, zone_id, record_name, record_type, address)
```

Comment:
Updating records cannot be done with `PUT /.../rrsets/...`, you must use one of the following endpoint:
- https://docs.hetzner.cloud/reference/cloud#zone-rrset-actions-set-records-of-an-rrset
- https://docs.hetzner.cloud/reference/cloud#zone-rrset-actions-add-records-to-an-rrset
- https://docs.hetzner.cloud/reference/cloud#zone-rrset-actions-remove-records-from-an-rrset

Also note, that those endpoint return "Actions", which describe async tasks that should be waited upon.